### PR TITLE
GEODE-5259: Remove class id from data serializable

### PR DESCRIPTION
**Pull request by mmartell**

### CODE REVIEW

**Comment by pivotal-jbarrett** [comment]
Can’t wait to take time to review this. Why can’t we have fully autonomous cars already!!

File: clicache/integration-test/DefaultCacheableN.cs [hunk 1]
```diff
@@ -156,29 +132,13 @@ public object CObject
 
     #region IDataSerializable Members
 
-    public Int32 ClassId
+    public String Type
```
Comment (pivotal-jbarrett) [inline clicache/integration-test/DefaultCacheableN.cs]: Now that the C++ doesn’t lookup this type anymore can’t we just used GetType on Object?

File: clicache/src/ManagedDataSerializableHandler.hpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+#include "begin_native.hpp"
+#include <SerializationRegistry.hpp>
+#include "end_native.hpp"
+
+#include "impl/CacheResolver.hpp"
+#include "impl/ManagedCacheableKey.hpp"
+#include "impl/SafeConvert.hpp"
+
+namespace apache {
+	namespace geode {
```
Comment (pivotal-jbarrett) [inline clicache/src/ManagedDataSerializableHandler.hpp]: Formatting? Looks like tabs?

File: clicache/integration-test/DefaultCacheableN.cs (hunk 1 above)
Comment (mmartell) [reply clicache/integration-test/DefaultCacheableN.cs]: Indeed we can. Good catch.

Done.

File: examples/dotnet/DataSerializableCs/Program.cs [hunk 3]
```diff
@@ -30,7 +30,7 @@ public static void Main(string[] args)
 
             Console.WriteLine("Registering for data serialization");
 
-            cache.TypeRegistry.RegisterType(Order.CreateDeserializable);
+            cache.TypeRegistry.RegisterType(Order.CreateDeserializable, 7);
```
Comment (echobravopapa) [inline examples/dotnet/DataSerializableCs/Program.cs]: What was wrong with 42? ;)

File: clicache/integration-test2/SerializationTests.cs [hunk 4]
```diff
@@ -24,108 +24,543 @@
...
+
+    public UInt64 ObjectSize
+    {
+      get
+      {
+        return (UInt32)(sizeof(Int32) + sizeof(Int64));
+      }
+    }
+
+    //public Int32 ClassId
```
[the remark this answers is not in the document]
Comment (mmartell) [reply clicache/integration-test2/SerializationTests.cs]: Good catch Michael.

File: clicache/integration-test2/SerializationTests.cs [hunk 5]
```diff
@@ -24,108 +24,543 @@
...
         private void putAndCheck(IRegion<object, object> region, object key, object value)
         {
-            region[key] = value;
-            Assert.Equal(value, region[key]);
+          region[key] = value;
+          var retrievedValue = region[key];
+          Assert.Equal(value, retrievedValue);
         }
 
+        private void putAndCheck2(IRegion<String, OtherType> region, String key, OtherType value)
```
[the remark this answers is not in the document]
Comment (mmartell) [reply clicache/integration-test2/SerializationTests.cs]: Nice catch.

File: cppcache/include/geode/DataSerializable.hpp [hunk 6]
```diff
@@ -48,14 +51,13 @@ class APACHE_GEODE_EXPORT DataSerializable : public virtual Serializable {
   virtual void fromData(DataInput& dataInput) = 0;
 
   /**
-   * @brief Return the classId of the instance being serialized.
+   * @brief Return the type name of the instance being serialized.
```
Comment (pivotal-jbarrett) [inline cppcache/include/geode/DataSerializable.hpp]: `std::type_info` not "type name"

File: cppcache/include/geode/internal/DSCode.hpp [hunk 7]
```diff
@@ -31,12 +31,10 @@ enum class DSCode : int32_t {
   FixedIDShort = 2,
   CacheableLinkedList = 10,
   Properties = 11,
-  CacheableToken = 14,  // because there's no equivalence in java
   PdxType = 17,  // internal hack to read pdxtype in c# layer, look usuage in
```
Comment (pivotal-jbarrett) [inline cppcache/include/geode/internal/DSCode.hpp]: I think `PdxType` can go away. It isn't really a `DSCode`.

File: cppcache/integration-test-2/CMakeLists.txt [hunk 8]
```diff
@@ -16,20 +16,21 @@
...
-  framework/Cluster.cpp
-  framework/Cluster.h
-  framework/GfshExecute.cpp
-  framework/GfshExecute.h
-  RegionPutGetAllTest.cpp
-  PdxInstanceTest.cpp
-  StructTest.cpp
-  EnableChunkHandlerThreadTest.cpp
-  )
+ ExampleTest.cpp
```
Comment (pivotal-jbarrett) [inline cppcache/integration-test-2/CMakeLists.txt]: 2 spaces

File: cppcache/integration-test-2/CMakeLists.txt [hunk 9]
```diff
@@ -38,39 +39,41 @@ target_compile_definitions(integration-test-2
...
-    GTest::GTest
-    GTest::Main
-    Boost::boost
-    Boost::system
-    Boost::log
-    Boost::filesystem
-  PRIVATE
-    _WarningsAsError
-    internal
+ PUBLIC
```
Comment (pivotal-jbarrett) [inline cppcache/integration-test-2/CMakeLists.txt]: 2 spaces

File: cppcache/src/CacheFactory.cpp [hunk 10]
```diff
@@ -100,23 +100,25 @@ Cache CacheFactory::create() const {
...
+  serializationRegistry->addDataSerializableFixedIdType(
       std::bind(TXCommitMessage::create, memberListForVersionStamp));
 
-  serializationRegistry->addType(
-      //TODO: This looks like the only thing to do here, but I'm not sure
-      static_cast<int64_t>(DSCode::PdxType),
-      std::bind(PdxType::CreateDeserializable, std::ref(*pdxTypeRegistry)));
+  serializationRegistry->setDataSerializablePrimitiveType(
+      // TODO: This looks like the only thing to do here, but I'm not sure
+      std::bind(PdxType::CreateDeserializable, std::ref(*pdxTypeRegistry)),
```
Comment (pivotal-jbarrett) [inline cppcache/src/CacheFactory.cpp]: This should be refactored out. It is not really a primitive type, it just pretends to be one. Give it the same treatment we did for `PdxTypeHandler` and `DataSerializableHandler`. In fact rename `PdxTypeHander` to `PdxSerializableHandler`, the current name is confusing with `PdxType`, which is the meta-data about PDX objects. Then implement this `PdxType` logic as `PdxTypeHandler`.

File: examples/dotnet/DataSerializableCs/Program.cs (hunk 3 above)
Comment (mmartell) [reply examples/dotnet/DataSerializableCs/Program.cs]: I like 7.

File: cppcache/include/geode/internal/DSCode.hpp (hunk 7 above)
Comment (mmartell) [reply cppcache/include/geode/internal/DSCode.hpp]: It's still used by `DistributedSystem::AppDomainInstanceInitialization` and `TcrMessage`. Will address in separate ticket: https://issues.apache.org/jira/browse/GEODE-5743

File: cppcache/src/CacheFactory.cpp (hunk 10 above)
Comment (mmartell) [reply cppcache/src/CacheFactory.cpp]: Will address in separate ticket: https://issues.apache.org/jira/browse/GEODE-5743

**Pull request merged**